// graphql-compose/lib/ListComposer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListComposer = void 0;

var _graphql = require("./graphql");

var _typeHelpers = require("./utils/typeHelpers");

var _NonNullComposer = require("./NonNullComposer");

class ListComposer {
  constructor(type) {
    this.ofType = type; // alive proper Flow type casting in autosuggestions for class with Generics

    /* :: return this; */
  }

  getType() {
    return new _graphql.GraphQLList(this.ofType.getType());
  }

  getTypeName() {
    return `[${this.ofType.getTypeName()}]`;
  }

  getUnwrappedTC() {
    let tc = this;

    while (!(0, _typeHelpers.isNamedTypeComposer)(tc)) {
      tc = tc.ofType;
    }

    return tc;
  }

  getTypePlural() {
    return new ListComposer(this);
  }

  getTypeNonNull() {
    return new _NonNullComposer.NonNullComposer(this);
  }
  /**
   * Clone this type to another SchemaComposer.
   * Also will be clonned all wrapped types.
   */


  cloneTo(anotherSchemaComposer, cloneMap = new Map()) {
    return new ListComposer(this.ofType.cloneTo(anotherSchemaComposer, cloneMap));
  }

}

exports.ListComposer = ListComposer;
