// remark-autolink-headings/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = attacher;

var _unistUtilVisit = _interopRequireDefault(require("unist-util-visit"));

var _extend = _interopRequireDefault(require("extend"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const behaviors = {
  prepend: 'unshift',
  append: 'push'
};
const contentDefaults = {
  type: 'element',
  tagName: 'span',
  properties: {
    className: ['icon', 'icon-link']
  }
};
const defaults = {
  behavior: 'prepend',
  content: contentDefaults
};
let deprecationWarningIssued = false;

function attacher(opts = {}) {
  let {
    linkProperties,
    behavior,
    content
  } = _objectSpread({}, defaults, {}, opts);

  let method;
  let hChildren; // NOTE: Remove in next major version

  if (opts.behaviour !== undefined) {
    if (!deprecationWarningIssued) {
      deprecationWarningIssued = true;
      console.warn('[remark-autolink-headings] Deprecation Warning: `behaviour` is a nonstandard option. Use `behavior` instead.');
    }

    behavior = opts.behaviour;
  }

  if (behavior === 'wrap') {
    method = wrap;
  } else {
    method = inject;
    hChildren = Array.isArray(content) ? content : [content];

    if (!linkProperties) {
      linkProperties = {
        ariaHidden: 'true'
      };
    }
  }

  return ast => (0, _unistUtilVisit.default)(ast, 'heading', visitor);

  function visitor(node) {
    const {
      data
    } = node;
    const id = data && data.hProperties && data.hProperties.id;

    if (id) {
      method(node, '#' + id);
    }
  }

  function inject(node, url) {
    node.children[behaviors[behavior]]({
      type: 'link',
      url,
      title: null,
      children: [],
      data: {
        hProperties: (0, _extend.default)(true, {}, linkProperties),
        hChildren: (0, _extend.default)(true, [], hChildren)
      }
    });
  }

  function wrap(node, url) {
    node.children = [{
      type: 'link',
      url,
      title: null,
      children: node.children,
      data: {
        hProperties: (0, _extend.default)(true, {}, linkProperties)
      }
    }];
  }
}

module.exports = exports.default;
